### PR TITLE
Separate Conditions for speech and silence

### PR DESCRIPTION
**Pull request by arifulhaque**
Changes made according to the comments of Ryan and Kevin.

### CODE REVIEW

**Comment by shkyang** [comment]
I have a question about the current VAD system:

how does it do in 'trimming' silence? If I gave it a mostly silent file (Irene and Forsad have several from Muscle), can it figure out where the silent portions are?

Arif, can you look into that and see what portion of the file is marked as silence?

**Comment by stansalvador** [comment]
I think we can't much more out of an energy-based VAD approach.
A follow-up story could be change the speech/non-speech model from an energy baed decision to using Speech/nonspeech outputs from a specially trained DNN.  (it can be largely the same as the normal AM's DNN but with ~2 output targets for speech and non-speech).

**Comment by arifulhaque** [comment]
"Minimum energy threshold" parameter is included to detect long initial silence.

File: cpp/config/configset.cpp [hunk 1]
```diff
@@ -37,6 +37,12 @@ namespace cobalt
         addDefaultParam("min_event_threshold", "300", \
                         "Length in ms of silence or speech to see before "
                         "event is created", INT_TYPE);
+        addDefaultParam("fe.vad_forgetting_factor", "0.9", \
+                        "Parameter controlling the weigth of the previous chunks "
```
Comment (kbasye27) [inline cpp/config/configset.cpp]: Typo - weigth -> weight

File: cpp/detector/kaldi_vad_detector.h [hunk 2]
```diff
@@ -44,31 +51,43 @@ namespace cobalt
...
                 int denCount = 0;
                 // moving window of frames
                 for (int windowedFrame = frame - context; windowedFrame <= frame + context; windowedFrame++)
                 {
-                    // accumulate stats when moving window within range
                     if (windowedFrame >= 0 && windowedFrame < totalFramesSeen)
                     {
                         denCount++;
-                        if (mLogEnergyAllFrames[windowedFrame] > energyThreshold)
+                        if (mFlagInSpeech == 0)
```
Comment (kbasye27) [inline cpp/detector/kaldi_vad_detector.h]: Since this is a bool, just use if (mFlagInSpeech) ...

File: cpp/config/configset.cpp [hunk 3]
```diff
@@ -37,6 +37,12 @@ namespace cobalt
...
+                        "in recursive calculation of mean energy", FLOAT_TYPE);
+        addDefaultParam("min_energy_threshold", "12.0", \
```
Comment (shkyang) [inline cpp/config/configset.cpp]: now we have both vad_energy_threshold, and min_energy_threshold? can we make the distinction clear in the names?

File: cpp/config/configset.cpp [hunk 4]
```diff
@@ -37,6 +37,12 @@ namespace cobalt
         addDefaultParam("min_event_threshold", "300", \
                         "Length in ms of silence or speech to see before "
                         "event is created", INT_TYPE);
+        addDefaultParam("fe.vad_forgetting_factor", "0.9", \
```
Comment (shkyang) [inline cpp/config/configset.cpp]: can this be called something else? forgetting_factor, its the scale of importance for previous events? I think there's an established name for it in CMVN, Stan should know what that is.

File: cpp/detector/kaldi_vad_detector.h [hunk 5]
```diff
@@ -9,8 +9,8 @@ namespace cobalt
     class KaldiVadDetector
     {
     public:
-        KaldiVadDetector(const kaldi::VadEnergyOptions &vadOptions)
-        : mVadOptions(vadOptions), mLogEnergySumAllFrames(0.0), mNumFramesAboveThreshold(0), mDenCount(0) {}
+        KaldiVadDetector(const kaldi::VadEnergyOptions &vadOptions, const float &forgettingFactor, const float &minEnergyThreshold)
```
Comment (shkyang) [inline cpp/detector/kaldi_vad_detector.h]: no need for reference on floats.

File: cpp/detector/kaldi_vad_detector.h [hunk 6]
```diff
@@ -9,8 +9,8 @@ namespace cobalt
...
+        : mVadOptions(vadOptions), mLogEnergySumAllFrames(0.0), mLogEnergyMeanPrevFrames(100000.0), mForgettingFactor(forgettingFactor), mMinEnergyThreshold(minEnergyThreshold), mFlagInSpeech(0) {}
```
Comment (shkyang) [inline cpp/detector/kaldi_vad_detector.h]: why 100000?

File: cpp/detector/kaldi_vad_detector.h [hunk 7]
```diff
@@ -30,11 +30,18 @@ namespace cobalt
             }
             kaldi::BaseFloat energyThreshold = mVadOptions.vad_energy_threshold;
             int totalFramesSeen = mLogEnergyAllFrames.size();
+            float logEnergyMeanCurrFrames = mLogEnergySumAllFrames / totalFramesSeen;
+            if (mLogEnergyMeanPrevFrames != 100000.0)
```
Comment (shkyang) [inline cpp/detector/kaldi_vad_detector.h]: hard coded numbers like (that aren't sentinel values like 0.0 or -1.0) make me nervous.

Comment (kbasye27) [inline cpp/detector/kaldi_vad_detector.h]: Would prefer {} after if (...) even for just one statement.

File: cpp/detector/kaldi_vad_detector.h [hunk 8]
```diff
@@ -44,31 +51,43 @@ namespace cobalt
...
+                        {
+                            if (mLogEnergyAllFrames[windowedFrame] < energyThreshold)
+                                numFramesBelowThreshold++;
                         }
                     }
-                 }
-                 const bool frameInSpeech = (numFramesAboveThreshold >= denCount * mVadOptions.vad_proportion_threshold);
-                 outputVoiced->push_back(frameInSpeech);
+                }
+                if (mFlagInSpeech == 0)
```
Comment (kbasye27) [inline cpp/detector/kaldi_vad_detector.h]: just if (mFlagInSpeech) and add {}s

File: cpp/detector/kaldi_vad_detector.h [hunk 9]
```diff
@@ -44,31 +51,43 @@ namespace cobalt
             {
                 int context = mVadOptions.vad_frames_context;
                 int numFramesAboveThreshold = 0;
+                int numFramesBelowThreshold = 0;
                 int denCount = 0;
```
Comment (kbasye27) [inline cpp/detector/kaldi_vad_detector.h]: Sorry, confused about this name - what is 'den' here?  density?

File: cpp/detector/kaldi_vad_detector.h (hunk 6 above)
Comment (kbasye27) [inline cpp/detector/kaldi_vad_detector.h]: How about -1.0 for this sentinel value?

File: cpp/detector/kaldi_vad_detector.h [hunk 10]
```diff
@@ -9,8 +9,8 @@ namespace cobalt
     class KaldiVadDetector
     {
     public:
-        KaldiVadDetector(const kaldi::VadEnergyOptions &vadOptions)
-        : mVadOptions(vadOptions), mLogEnergySumAllFrames(0.0), mNumFramesAboveThreshold(0), mDenCount(0) {}
+        KaldiVadDetector(const kaldi::VadEnergyOptions &vadOptions, const float &forgettingFactor, const float &minEnergyThreshold)
+        : mVadOptions(vadOptions), mLogEnergySumAllFrames(0.0), mLogEnergyMeanPrevFrames(100000.0), mForgettingFactor(forgettingFactor), mMinEnergyThreshold(minEnergyThreshold), mFlagInSpeech(0) {}
 
         void computeVadEnergy(const kaldi::MatrixBase<kaldi::BaseFloat> &inputFeatures,
                               std::vector<bool> *outputVoiced)
```
Comment (kbasye27) [inline cpp/detector/kaldi_vad_detector.h]: Just below this I think we should add a check that NumRows() (or numFrames) is also not 0.  Either just return like we do for NumCols() or maybe just an assert - but right now it looks like we'll divide by 0 later on and put a NaN into logEnergyMeanCurrFrames.

File: cpp/detector/kaldi_vad_detector.h (hunk 5 above)
Comment (kbasye27) [inline cpp/detector/kaldi_vad_detector.h]: No need for const on floats, either.

File: cpp/detector/kaldi_vad_detector.h [hunk 11]
```diff
@@ -30,11 +30,18 @@ namespace cobalt
             }
             kaldi::BaseFloat energyThreshold = mVadOptions.vad_energy_threshold;
             int totalFramesSeen = mLogEnergyAllFrames.size();
+            float logEnergyMeanCurrFrames = mLogEnergySumAllFrames / totalFramesSeen;
+            if (mLogEnergyMeanPrevFrames != 100000.0)
+                logEnergyMeanCurrFrames = mForgettingFactor*mLogEnergyMeanPrevFrames + (1-
```
Comment (kbasye27) [inline cpp/detector/kaldi_vad_detector.h]: Maybe add a KALDI_ASSERT like the one below that mForgettingFactor is >- 0.0 and <= 1.0 ?

**Comment by kbasye27** [comment]
OK, I've made a handful of suggestions, mostly about style.  The fundamental algorithmic changes look good to me.

File: cpp/detector/kaldi_vad_detector.h (hunk 9 above)
Comment (arifulhaque) [inline cpp/detector/kaldi_vad_detector.h]: I am also not sure. However, I think it comes from 'denominator'.

**Comment by arifulhaque** [comment]
Changes made according to the comments of Kevin and Ken Basye. Shall I merge it with the master?

**Comment by arifulhaque** [comment]
It is long overdue. I made all the changes according to the comments of Kevin and Ken Basye. Shall I merge it with tmaster?

**Pull request merged**